### DISCUSSION TITLE
[BUG] User deletetion does not work

### DISCUSSION THREAD

**Discussion by al13nus** [immich-app/immich#7963] [Q&A] [answered]
### The bug

https://github.com/immich-app/immich/issues/1086
https://github.com/immich-app/immich/pull/1090

Hey, nearly the same issue here: I delete a user and waited for more then 3 weeks. I don't reminde the date but it was something like 3. march. After this period the user is still shown but I cannot remove him or create a new one with this email:

![grafik](https://github.com/immich-app/immich/assets/19974381/9fe92d84-4370-450c-911d-abf78e659c6b)

### The OS that Immich Server is running on

Ubuntu 22

### Version of Immich Server

1.98.2

### Version of Immich Mobile App

1.98.2

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
standard
```

### Your .env content

```Shell
in docker stack, standard
```

### Reproduction steps

```bash
1. click to delete user
2. wait the period where he should be deleted
3. check if really deleted
...
```

```[tasklist]
### Tasks
```

**Comment by jrasm91** [comment]
It is possible the user failed to delete because of a bug. Can you check the immich-microservice logs to see if there is any indication there as to why it failed? The job to delete a user should run ever night so you shouldn't have to go too far back.

**Comment by monkeyotg** [comment]
same here, but the data, folder on the disk from these accounts are deleted, just the user-recorder is still there.
I think perhaps it is a feature to let you check which account is deleted. 😅

![Snipaste_2024-03-13_22-45-01](https://github.com/immich-app/immich/assets/8792193/80661ee1-acb8-47f8-a624-44b71e6e58b1)

**Comment by samholton** [comment]
[v1.98.1](https://github.com/immich-app/immich/releases/tag/v1.98.1) fixed an issue deleting users that still had active shared links - so that shouldn't be an issue in v1.98.2

The log messages you are looking for are:

```
Deleting user: <user uuid>

Removing user from filesystem:

Removing user from database:
```

If you can check the database, the output of the following query could be helpful as well

```sql
select "email", "deletedAt", "status"
from users
where email = 'user@email.com';
```

Also, just to confirm, you haven't changed the `DELETE DELAY` option under `User Settings` on the Admin System Settings?

**Comment by monkeyotg** [comment]
I've changed the DELETE DELAY option under User Settings. Because I am sure deleting these accounts is safe, so I've changed it to 1. And found out it didn't work, so changed it back to 7 days.

**Comment by jrasm91** [comment]
Can you check the immich-microservices logs for any errors?

**Comment by al13nus** [comment, marked as answer]
Looked into logs:

```
[Nest] 7  - 03/14/2024, 12:00:00 AM   ERROR [JobService] Unable to run job handler (backgroundTask/user-deletion): Error: EBUSY: resource busy or locked, rmdir 'upload/library/mariia'
[Nest] 7  - 03/14/2024, 12:00:00 AM   ERROR [JobService] Error: EBUSY: resource busy or locked, rmdir 'upload/library/mariia'
```

Mounting to docker:

```
/mnt/tank/Media/UserData/mariia/Photos | /usr/src/app/upload/library/mariia
``` 

I deleted now the folder manually and will wait 24h for final cleanup.

**Reply by al13nus** [reply]
After removing the folder manually, the user disappeared next day.

**Comment by monkeyotg** [comment]
> Can you check the immich-microservices logs for any errors?

It was a long time ago, so there is no log of that day that I can find now.
Like I said, at the time I thought it was a feature.

**Comment by jrasm91** [comment]
> > Can you check the immich-microservices logs for any errors?
> 
> It was a long time ago, so there is no log of that day that I can find now. Like I said, at the time I thought it was a feature.

If it hasn't been deleted yet, but it is showing red like that, which means it is "soft deleted", it will try to complete the deletion every night, so there should be errors everyday.

**Comment by monkeyotg** [comment]
Somehow I think I've deleted it once. (But I don’t remember it at all.) Now the oldest log is 2024-03-11 and there are some errors. But I don't know if they are related. (Sorry, I don't know coding.)

![image](https://github.com/immich-app/immich/assets/8792193/184f0c3c-0605-4865-b8fc-bdee91f6b119)

these errors show themselves about 4 times a day
[immich_microservices_log.txt](https://github.com/immich-app/immich/files/14606161/immich_microservices_log.txt)

**Reply by jrasm91** [reply]
Yeah none of those are relevant. Do you not have a user that is pending deletion still or no?

**Reply by monkeyotg** [reply]
This is how it looks like under User Management now

![image](https://github.com/immich-app/immich/assets/8792193/a2f8e6f7-b255-4734-a975-02d19138ea49)

**Reply by jrasm91** [reply]
Did you force delete those users?

**Reply by monkeyotg** [reply]
What does "force delete" mean?

**Reply by jrasm91** [reply]
Did you delete these users more than a week ago? You just did the normal process?

**Reply by monkeyotg** [reply]
Yes, more than a week (at least one of them). But the exact process I can't recall. But I am sure, I just clicked the buttons shown on the interface.

---
Update:
I've made a check again and noticed that the program's automatic folder cleanup is not performed thoroughly. Some corpses were left behind. Now I've removed them all.

**Reply by monkeyotg** [reply]
A quick feedback:
The accounts are now gone.
Seems like the result of the "Task: auto-delete after 7 days", but there's a lot of weirdness in it.
I'll sort it out and add more.

**Reply by jrasm91** [reply]
If it failed to delete stuff there would be an error in the log, which is why I kept asking for them.

**Comment by Altycoder** [comment]


**Reply by Altycoder** [reply]
I reduced the number of days to 3 and last night got the following errors in the microservices log:

```
[Nest] 7  - 05/23/2024, 12:00:00 AM     LOG [ImmichMicroservices] [UserService] Deleting user: 91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:00 AM    WARN [ImmichMicroservices] [UserService] Removing user from filesystem: upload/library/91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:00 AM    WARN [ImmichMicroservices] [UserService] Removing user from filesystem: upload/upload/91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:00 AM    WARN [ImmichMicroservices] [UserService] Removing user from filesystem: upload/profile/91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:00 AM    WARN [ImmichMicroservices] [UserService] Removing user from filesystem: upload/thumbs/91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:07 AM    WARN [ImmichMicroservices] [UserService] Removing user from filesystem: upload/encoded-video/91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:07 AM    WARN [ImmichMicroservices] [UserService] Removing user from database: 91fb45dd-f8c6-474e-94a8-25960f31b3b3
[Nest] 7  - 05/23/2024, 12:00:22 AM   ERROR [ImmichMicroservices] [JobService] Unable to run job handler (backgroundTask/user-deletion): QueryFailedError: update or delete on table "assets" violates foreign key constraint "FK_91704e101438fd0653f582426dc" on table "asset_stack"
    at async UserService.handleUserDelete (/usr/src/app/dist/services/user.service.js:179:9)
```

**Comment by NicholasFlamy** [comment]
Closing this as it seems this has been dealt with.

**Comment by Altycoder** [comment]
Can't someone please respond to my similar bug please: https://github.com/immich-app/immich/issues/9693
